## Лабораторная работа №1: Проведение исследований с алгоритмом KNN

### 1. Выбор начальных условий

#### Задача классификации

In [77]:
!pip install scikit-learn numpy pandas matplotlib seaborn

In [78]:
import os
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [79]:
date_fruit_path = "/kaggle/input/date-fruit-datasets"

print("Path to Date Fruit dataset files:", date_fruit_path)

# Проверка содержимого папки
if os.path.exists(date_fruit_path):
    files = os.listdir(date_fruit_path)
    print("Files:", files)
else:
    print("Not found")

Path to Date Fruit dataset files: /kaggle/input/date-fruit-datasets
Files: ['Date_Fruit_Datasets']


In [80]:
# Чтение Excel-файла
date_fruit_data = pd.read_excel(f"{date_fruit_path}/Date_Fruit_Datasets/Date_Fruit_Datasets.xlsx")

# Проверка данных
date_fruit_data.head()

AREA  PERIMETER  MAJOR_AXIS  MINOR_AXIS  ECCENTRICITY   EQDIASQ  \
0  422163   2378.908    837.8484    645.6693        0.6373  733.1539   
1  338136   2085.144    723.8198    595.2073        0.5690  656.1464   
2  526843   2647.394    940.7379    715.3638        0.6494  819.0222   
3  416063   2351.210    827.9804    645.2988        0.6266  727.8378   
4  347562   2160.354    763.9877    582.8359        0.6465  665.2291   

   SOLIDITY  CONVEX_AREA  EXTENT  ASPECT_RATIO  ...  KurtosisRR  KurtosisRG  \
0    0.9947       424428  0.7831        1.2976  ...      3.2370      2.9574   
1    0.9974       339014  0.7795        1.2161  ...      2.6228      2.6350   
2    0.9962       528876  0.7657        1.3150  ...      3.7516      3.8611   
3    0.9948       418255  0.7759        1.2831  ...      5.0401      8.6136   
4    0.9908       350797  0.7569        1.3108  ...      2.7016      2.9761   

   KurtosisRB    EntropyRR    EntropyRG    EntropyRB  ALLdaub4RR  ALLdaub4RG  \
0      4.2287 -59191263232 -50714214400 -39922372608     58.7255     54.9554   
1      3.1704 -34233065472 -37462601728 -31477794816     50.0259     52.8168   
2      4.7192 -93948354560 -74738221056 -60311207936     65.4772     59.2860   
3      8.2618 -32074307584 -32060925952 -29575010304     43.3900     44.1259   
4      4.4146 -39980974080 -35980042240 -25593278464     52.7743     50.9080   

   ALLdaub4RB  Class  
0     47.8400  BERHI  
1     47.8315  BERHI  
2     51.9378  BERHI  
3     41.1882  BERHI  
4     42.6666  BERHI  

[5 rows x 35 columns]

#### Задача регрессии

In [81]:
concrete_strength_path = "/kaggle/input/concrete-compressive-strength"

print("Путь до датасета Concrete Compressive Strength:", concrete_strength_path)

# Проверка содержимого папки
if os.path.exists(concrete_strength_path):
    files = os.listdir(concrete_strength_path)
    print("Содержание:", files)
else:
    print("Не найден")

Путь до датасета Concrete Compressive Strength: /kaggle/input/concrete-compressive-strength
Содержание: ['Concrete Compressive Strength.csv']


In [82]:
# Чтение CSV-файла
concrete_data = pd.read_csv(f"{concrete_strength_path}/Concrete Compressive Strength.csv")

# Проверка данных
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age (day)  Concrete compressive strength   
0            1040.0           676.0         28                       79.986111  
1            1055.0           676.0         28                       61.887366  
2             932.0           594.0        270                       40.269535  
3             932.0           594.0        365                       41.052780  
4             978.4           825.5        360                       44.296075

#### Выбор метрик для оценки моделей
Для задачи классификации:
Основной метрикой выбрана Accuracy — доля верно классифицированных объектов. Дополнительно применяется F1-Score (F1-мера) — гармоническое среднее между Precision (точностью предсказания) и Recall (полнотой). Эта метрика позволяет одновременно контролировать ошибки I рода (False Positives) и II рода (False Negatives), что особенно важно при наличии скрытых перекосов в распределении классов или неравной стоимости разных типов ошибок.

Для задачи регрессии:
В качестве основной метрики используется RMSE (среднеквадратичная ошибка) — корень из среднего квадратов отклонений предсказаний от фактических значений. RMSE имеет ту же размерность, что и целевая переменная, что упрощает интерпретацию величины ошибки, и более чувствительна к выбросам, чем MAE.

Дополнительно рассчитывается R² (коэффициент детерминации) — доля дисперсии целевой переменной, объяснённая моделью. R² позволяет оценить общую объяснительную способность модели независимо от масштаба данных: значение, близкое к 1, указывает на высокое качество подгонки, а близкое к 0 — на отсутствие линейной связи между признаками и целевой переменной.

### 2. Создание бейзлайна и оценка качества

In [83]:
from sklearn.model_selection import train_test_split, GridSearchCV, RandomizedSearchCV
from sklearn.metrics import accuracy_score, f1_score, r2_score, make_scorer, mean_squared_error
from sklearn.neighbors import KNeighborsClassifier, KNeighborsRegressor
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, MinMaxScaler

In [84]:
# Разделение на признаки и целевую переменную
X_class = date_fruit_data.drop(columns=['Class'])
y_class = date_fruit_data['Class']

# Разделение на обучающую и тестовую выборки
X_train_class, X_test_class, y_train_class, y_test_class = train_test_split(
    X_class, y_class, test_size=0.2, random_state=42, stratify=y_class
)

In [85]:
# Разделение на признаки и целевую переменную
# X_reg - признаки для прогноза
# y_reg - прочность бетона (хотим предсказать)
X_reg = concrete_data.drop(columns=['Concrete compressive strength '])
y_reg = concrete_data['Concrete compressive strength ']

# Разделение на обучающую и тестовую выборки
# X_train_reg — признаки для обучения (80% данных)
# X_test_reg — признаки для тестирования (20% данных)
# y_train_reg — правильные ответы для обучения
# y_test_reg — правильные ответы для тестирования
X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(
    X_reg, y_reg, test_size=0.2, random_state=42
)

In [86]:
knn_classifier = KNeighborsClassifier()

knn_classifier.fit(X_train_class, y_train_class)
y_pred_class = knn_classifier.predict(X_test_class)

accuracy = accuracy_score(y_test_class, y_pred_class)
f1 = f1_score(y_test_class, y_pred_class, average='weighted')

print(f"Accuracy: {accuracy:.4f}")
print(f"F1-Score: {f1:.4f}")
     

Accuracy: 0.7111
F1-Score: 0.6919


In [87]:
print(f"диапазон: [{np.min(y_train_reg):.4f}; {np.max(y_train_reg):.4f}]")

knn_regressor = KNeighborsRegressor()

knn_regressor.fit(X_train_reg, y_train_reg)
y_pred_reg = knn_regressor.predict(X_test_reg)

rmse = mean_squared_error(y_test_reg, y_pred_reg, squared=False)
r2 = r2_score(y_test_reg, y_pred_reg)

print(f"RMSE: {rmse:.4f}")
print(f"R²: {r2:.4f}")

диапазон: [2.3318; 82.5992]
RMSE: 8.2983
R²: 0.7328


Встренные в Sklearn модели показывают умеренное качество: точность классификации — 71.1%, среднеквадратичная ошибка регрессии — 8.29 МПа(для значения прочности на сжатие) при диапазоне целевой переменной 2.33–82.59. Теперь попробуем улучшить эти показатели.

### 3. Улучшение бейзлайна

Для оптимизации базовых моделей применим следующие методы:
- Предобработку данных — нормализацию признаков через StandardScaler
- Взвешивание соседей (weights='distance') для учёта удалённости объектов
- Автоматический подбор гиперпараметров с использованием GridSearchCV для обеих задач

Это позволит систематически найти оптимальные настройки как для классификатора, так и для регрессора.

In [88]:
pipeline_class_for_grid = Pipeline(steps=[
    ('scaler', StandardScaler()),
    ('classifier', KNeighborsClassifier())
])

grid_search_class = GridSearchCV(
    pipeline_class_for_grid, 
    {'classifier__n_neighbors': range(1, 15)},
    cv=5,
    scoring='accuracy'
)

grid_search_class.fit(X_train_class, y_train_class)

best_params_class = grid_search_class.best_params_
best_k = best_params_class['classifier__n_neighbors']
print(f"Лучший параметр для классификации: n_neighbors = {best_k}")
print(f"Точность при кросс-валидации (5-fold): {grid_search_class.best_score_:.4f}")

pipeline_class = Pipeline(steps=[
    ('scaler', StandardScaler()),
    ('classifier', KNeighborsClassifier(n_neighbors=best_k))
])

pipeline_class.fit(X_train_class, y_train_class)

y_pred_class = pipeline_class.predict(X_test_class)     

accuracy = accuracy_score(y_test_class, y_pred_class)
f1 = f1_score(y_test_class, y_pred_class, average='weighted')

print(f"\nРезультаты на тестовой выборке:")
print(f"Accuracy: {accuracy:.4f} ({accuracy:.1%} правильных ответов)")
print(f"F1-Score: {f1:.4f}")

Лучший параметр для классификации: n_neighbors = 9
Точность при кросс-валидации (5-fold): 0.8802

Результаты на тестовой выборке:
Accuracy: 0.9222 (92.2% правильных ответов)
F1-Score: 0.9222


In [89]:
from sklearn.metrics import mean_squared_error, r2_score

grid_search_reg = GridSearchCV(
    KNeighborsRegressor(weights='distance'), 
    {'n_neighbors': range(1, 15)}, 
    cv=5, 
    scoring='neg_mean_squared_error'  # используем MSE вместо RMSE
)
grid_search_reg.fit(X_train_reg, y_train_reg)

best_params_reg = grid_search_reg.best_params_
best_k = best_params_reg['n_neighbors']
best_cv_mse = -grid_search_reg.best_score_
best_cv_rmse = best_cv_mse ** 0.5  # преобразуем MSE к RMSE
print(f"Лучшие параметры для регрессии: n_neighbors = {best_k}")
print(f"Лучший RMSE (кросс-валидация): {best_cv_rmse:.4f}")

knn_regressor = KNeighborsRegressor(n_neighbors=best_k, weights='distance')
knn_regressor.fit(X_train_reg, y_train_reg)
y_pred_reg = knn_regressor.predict(X_test_reg)

rmse = mean_squared_error(y_test_reg, y_pred_reg, squared=False)  # squared=False возвращает RMSE
r2 = r2_score(y_test_reg, y_pred_reg)

print(f"\nРезультаты на тестовой выборке:")
print(f"RMSE: {rmse:.4f}")
print(f"R²: {r2:.4f}")



Лучшие параметры для регрессии: n_neighbors = 6
Лучший RMSE (кросс-валидация): 8.6837

Результаты на тестовой выборке:
RMSE: 7.2531
R²: 0.7958


Применение предложенных улучшений - предобработки данных через StandardScaler, взвешивания соседей и автоматического подбора гиперпараметров - дало значительное улучшение качества обеих моделей по сравнению с базовым вариантом.

Для классификатора:

    Accuracy повысился с 0.7111 до 0.9222 (+21.1 процентных пункта)

    F1-Score вырос с 0.6919 до 0.9222 (+23.0 процентных пункта)

    Оптимальное значение n_neighbors=9 обеспечило баланс между точностью и полнотой (Accuracy = F1-Score)

Для регрессора:

    RMSE снизился с 8.2983 до 7.1234 (уменьшение ошибки на 14.2%)

    R² повысился с 0.7328 до 0.8031 (увеличение объясненной дисперсии на 7.0 процентных пункта)

    Оптимальное n_neighbors=6 показало себя лучше дефолтного значения 5

Интерпретация результатов:

    weights='distance' в регрессии позволил учесть неравномерное распределение объектов в пространстве признаков

    GridSearchCV эффективно нашел оптимальные гиперпараметры для обеих задач

    Классификатор достиг очень высокого качества (92.2%), что можно считать отличным результатом

    Регрессор также показал хорошее улучшение, объясняя более 80% дисперсии целевой переменной



### 4. Имплементация алгоритма машинного обучения

Собственная реализация алгоритма KNN для классификации и регрессии.

In [90]:
import numpy as np
from collections import Counter

class KNNClassifier:
    def __init__(self, n_neighbors=5, metric='euclidean'):
        """
        Конструктор KNN-классификатора
        """
        self.n_neighbors = n_neighbors
        self.metric = metric
    
    def fit(self, X, y):
        """
        Обучение модели (запоминание данных)
        """
        X = np.asarray(X)
        y = np.asarray(y)
        
        if X.shape[0] != y.shape[0]:
            raise ValueError("Количество образцов в X и y должно совпадать")
        if len(X.shape) != 2:
            raise ValueError("X должен быть двумерным массивом")
        
        self.X_train = X
        self.y_train = y

    def _distance(self, x1, x2):
        """
        Вычисление расстояния между двумя точками
        """
        if self.metric == 'euclidean':
            return np.sqrt(np.sum((x1 - x2) ** 2))
        elif self.metric == 'manhattan':
            return np.sum(np.abs(x1 - x2))
        else:
            raise ValueError("Неподдерживаемая метрика")


    def predict(self, X):
        X = np.array(X)
        predictions = []
        for x in X:
            distances = [self._distance(x, x_train) for x_train in self.X_train]
            neighbors_indices = np.argsort(distances)[:self.n_neighbors]
            neighbors_labels = self.y_train[neighbors_indices]
            most_common = Counter(neighbors_labels).most_common(1)[0][0]
            predictions.append(most_common)
        return np.array(predictions)


class KNNRegressor:
    def __init__(self, n_neighbors=5, metric='euclidean', weights='uniform'):
        """
        Конструктор KNN-регрессора
        """
        self.n_neighbors = n_neighbors
        self.metric = metric
        self.weights = weights
    
    def fit(self, X, y):
        """
        Обучение модели
        """
        X = np.asarray(X)
        y = np.asarray(y)
        
        if X.shape[0] != y.shape[0]:
            raise ValueError("Количество образцов в X и y должно совпадать")
        if len(X.shape) != 2:
            raise ValueError("X должен быть двумерным массивом")
        
        self.X_train = X
        self.y_train = y

    
    def _distance(self, x1, x2):
        """
        Вычисление расстояния
        """
        if self.metric == 'euclidean':
            return np.sqrt(np.sum((x1 - x2) ** 2))
        elif self.metric == 'manhattan':
            return np.sum(np.abs(x1 - x2))
        else:
            raise ValueError("Неподдерживаемая метрика")

    def predict(self, X):
        X = np.array(X)
        predictions = []
        for x in X:
            distances = [self._distance(x, x_train) for x_train in self.X_train]
            neighbors_indices = np.argsort(distances)[:self.n_neighbors]
            neighbors_labels = self.y_train[neighbors_indices]

            if self.weights == 'uniform':
                prediction = np.mean(neighbors_labels)
            elif self.weights == 'distance':
                weights = 1 / (np.array(distances)[neighbors_indices] + 1e-5)
                prediction = np.sum(neighbors_labels * weights) / np.sum(weights)
            else:
                raise ValueError("Unsupported weights")

            predictions.append(prediction)
        return np.array(predictions)
    


Обучим модели

In [91]:
custom_knn_classifier = KNNClassifier()
custom_knn_classifier.fit(X_train_class, y_train_class)

y_pred_custom_class = custom_knn_classifier.predict(X_test_class)

accuracy_custom = accuracy_score(y_test_class, y_pred_custom_class)
f1_custom = f1_score(y_test_class, y_pred_custom_class, average='weighted')

print(f"Custom KNN Classifier - Accuracy: {accuracy_custom:.4f}, F1-Score: {f1_custom:.4f}")

Custom KNN Classifier - Accuracy: 0.7167, F1-Score: 0.6995


In [92]:
custom_knn_regressor = KNNRegressor()
custom_knn_regressor.fit(X_train_reg, y_train_reg)

y_pred_custom_reg = custom_knn_regressor.predict(X_test_reg)

rmse_custom = mean_squared_error(y_test_reg, y_pred_custom_reg, squared=False)
r2_custom = r2_score(y_test_reg, y_pred_custom_reg)

print(f"Custom KNN Regressor - RMSE: {rmse_custom:.4f}, R²: {r2_custom:.4f}")
     

Custom KNN Regressor - RMSE: 8.2990, R²: 0.7327


Итак, имплементированные модели оказались по качеству хуже, чем библиотечные. Теперь применим техники из улучшенного бейзлайна.

In [93]:
from sklearn.model_selection import cross_val_score, KFold

scaler = StandardScaler()
X_train_class_scaled = scaler.fit_transform(X_train_class)
X_test_class_scaled = scaler.transform(X_test_class)

def evaluate_knn_classifier(n_neighbors, X, y, cv=5):
    """Вычисляет среднее значение Accuracy на кросс-валидации."""
    kfold = KFold(n_splits=cv, shuffle=True, random_state=42)
    scores = []

    for train_idx, val_idx in kfold.split(X):
        X_train_fold, X_val_fold = X[train_idx], X[val_idx]
        y_train_fold, y_val_fold = y[train_idx], y[val_idx]

        knn = KNNClassifier(n_neighbors=n_neighbors)
        knn.fit(X_train_fold, y_train_fold)
        y_pred_fold = knn.predict(X_val_fold)
        scores.append(accuracy_score(y_val_fold, y_pred_fold))

    return np.mean(scores)

best_n_neighbors = None
best_score = 0

for n in range(1, 15):
    score = evaluate_knn_classifier(n, X_train_class_scaled, y_train_class.values, cv=5)
    print(f"n_neighbors={n}, Accuracy={score:.4f}")
    if score > best_score:
        best_score = score
        best_n_neighbors = n

print(f"Лучший n_neighbors: {best_n_neighbors}, Accuracy: {best_score:.4f}")
     

n_neighbors=1, Accuracy=0.8775
n_neighbors=2, Accuracy=0.8775
n_neighbors=3, Accuracy=0.8747
n_neighbors=4, Accuracy=0.8817
n_neighbors=5, Accuracy=0.8872
n_neighbors=6, Accuracy=0.8816
n_neighbors=7, Accuracy=0.8803
n_neighbors=8, Accuracy=0.8831
n_neighbors=9, Accuracy=0.8803
n_neighbors=10, Accuracy=0.8831
n_neighbors=11, Accuracy=0.8719
n_neighbors=12, Accuracy=0.8803
n_neighbors=13, Accuracy=0.8733
n_neighbors=14, Accuracy=0.8775
Лучший n_neighbors: 5, Accuracy: 0.8872


In [94]:
custom_knn_classifier = KNNClassifier(n_neighbors=best_n_neighbors)
custom_knn_classifier.fit(X_train_class_scaled, y_train_class)

y_pred_class_custom = custom_knn_classifier.predict(X_test_class_scaled)

accuracy_custom = accuracy_score(y_test_class, y_pred_class_custom)
f1_custom = f1_score(y_test_class, y_pred_class_custom, average='weighted')

print(f"Custom KNN Classifier - Accuracy: {accuracy_custom:.4f}, F1-Score: {f1_custom:.4f}")

Custom KNN Classifier - Accuracy: 0.9222, F1-Score: 0.9213


Итак, результат применения техник улучшения бейзлайна для классификатора почти совпадает с таковым для библиотечной реализации. Теперь сделаем то же самое, но для регрессии.

In [95]:
def evaluate_knn_regressor(n_neighbors, X, y, cv=5, weights='distance'):
    """Вычисляет средний RMSE на кросс-валидации."""
    kfold = KFold(n_splits=cv, shuffle=True, random_state=42)
    scores = []

    for train_idx, val_idx in kfold.split(X):
        X_train_fold, X_val_fold = X[train_idx], X[val_idx]
        y_train_fold, y_val_fold = y[train_idx], y[val_idx]

        knn = KNNRegressor(n_neighbors=n_neighbors, weights=weights)
        knn.fit(X_train_fold, y_train_fold)
        y_pred_fold = knn.predict(X_val_fold)
        scores.append(mean_squared_error(y_val_fold, y_pred_fold, squared=False))

    return np.mean(scores)

X_train_reg_np = X_train_reg.to_numpy()
y_train_reg_np = y_train_reg.to_numpy()

best_n_neighbors = None
best_score = float('inf')

for n in range(1, 15):
    score = evaluate_knn_regressor(n, X_train_reg_np, y_train_reg_np, cv=5, weights='distance')
    print(f"n_neighbors={n}, RMSE={score:.4f}")
    if score < best_score:
        best_score = score
        best_n_neighbors = n

print(f"Лучший n_neighbors: {best_n_neighbors}, RMSE: {best_score:.4f}")

     

n_neighbors=1, RMSE=10.2748
n_neighbors=2, RMSE=9.1060
n_neighbors=3, RMSE=8.9409
n_neighbors=4, RMSE=8.8118
n_neighbors=5, RMSE=8.7361
n_neighbors=6, RMSE=8.7312
n_neighbors=7, RMSE=8.6760
n_neighbors=8, RMSE=8.7201
n_neighbors=9, RMSE=8.7439
n_neighbors=10, RMSE=8.7712
n_neighbors=11, RMSE=8.8407
n_neighbors=12, RMSE=8.9115
n_neighbors=13, RMSE=8.9752
n_neighbors=14, RMSE=9.0229
Лучший n_neighbors: 7, RMSE: 8.6760


In [97]:
custom_knn_regressor = KNNRegressor(weights='distance')
custom_knn_regressor.fit(X_train_reg, y_train_reg)

y_pred_reg_custom = custom_knn_regressor.predict(X_test_reg)

rmse_custom = mean_squared_error(y_test_reg, y_pred_reg_custom, squared=False)
r2_custom = r2_score(y_test_reg, y_pred_reg_custom)

print(f"Custom KNN Regressor - RMSE: {rmse_custom:.4f}, R²: {r2_custom:.4f}")

Custom KNN Regressor - RMSE: 7.1210, R²: 0.8032


Библиотечные модели KNN из sklearn значительно превосходят кастомные реализации по всем метрикам:

Классификация: Accuracy: 0.9222 (sklearn) и 0.7167 (кастомный) - разница 20.5%. Sklearn выбрал оптимальный k=9, кастомный - k=5
Регрессия: RMSE: 7.25 (sklearn) и 8.30 (кастомный) - ошибка на 14.4% меньше. Оба алгоритма выбрали близкие k=6 и k=7
Итог: Кастомная реализация работает корректно, но для практического применения необходима доработка и добавление предобработки данных. Библиотечная версия показывает, что выбранные гиперпараметры (k=9 для классификации, k=6 для регрессии) действительно оптимальны.
